make `ResultMetadata` lifetime-generic

wprzytula opened the pull request:
Ref: #462 

This is a necessary step towards non-allocating (in fact less-allocating) result metadata deserialization.

# What's done

- `ColumnType` is made ownership-generic; there will be two distinct functions to deserialize owned and borrowed `ColumnType`;
- `ColumnSpec` is made ownership-generic; in a follow-up PR, there will be a distinct function to deserialize borrowed `Vec<ColumnSpec<'frame>>`.
- `ResultMetadata` is made ownership-generic;
- `deser_table_spec` no longer allocates; it's up to `deser_col_specs` to allocate or not, depending on desired ownership of deserialized metadata;
- `deser_col_specs` verifies that table specs are the same for all columns. We assume it's true, but the CQL protocol does not enforce it. With that check, we will be able to hold only one table spec per query, not one per column; this will reduce memory footprint. TODO in a follow-up.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~~[] I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- ~~[ ] I have provided docstrings for the public items that I want to introduce.~~
- ~~[ ] I have adjusted the documentation in `./docs/source/`.~~
- ~~[ ] I added appropriate `Fixes:` annotations to PR description.~~

muzarski commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -620,7 +621,7 @@ mod tests {
             .into_iter()
             .enumerate()
-            .map(|(i, typ)| ColumnSpec::new_for_test(table_spec.clone(), format!("col_{}", i), typ))
+            .map(|(i, typ)| ColumnSpec::owned(format!("col_{}", i), typ, table_spec.clone()))
```
Why we decided not to use `new_for_test` here anymore? I actually like `ColumnSpec::owned` here more, since we are explicit about creating an owned type.

muzarski replied:
I think we should remove `new_for_test` - I checked out and don't see any usages of this function anymore (replaced with `CS::owned` and `CS::borrowed` calls in tests).

muzarski commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -536,12 +536,12 @@ impl<'a> ColumnSpec<'a> {
-pub struct ResultMetadata {
+pub struct ResultMetadata<'a> {
     col_count: usize,
-    pub col_specs: Vec<ColumnSpec<'static>>,
```
nit: `pub` removal should belong to previous commit

muzarski replied:
commit `result: make ResultMetadata lifetime generic` - `ResultMetadata` struct

muzarski commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -619,14 +619,30 @@ pub enum Result {
-        .to_owned();
-    Ok(TableSpec::owned(ks_name, table_name))
+// Only allocates a new `TableSpec` if one is not yet given.
+fn deser_table_spec<'frame: 'spec, 'spec>(
```
Can we close https://github.com/scylladb/scylla-rust-driver/pull/956 then?

muzarski replied:
If we are going to store a borrowed `TableSpec` in each `ColumnSpec`, then I don't think we need to remove it from `ColumnSpec`.

wprzytula replied:
This is not the final state. In a follow-up of the deserialization refactor, I'm going to decouple `TableSpec` from `ColumnSpec`. The reason for doing this in a follow-up is to prove that such change won't break the public API, so an opposite change won't, too, which is what @Lorak-mmk believes is important in case Scylla/Cassandra once introduce joins to CQL.

muzarski replied:
Ok,, makes sense

muzarski commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -730,21 +746,30 @@ fn mk_col_spec_parse_error(
+        col_specs.push(ColumnSpec::owned(
+            name,
+            typ,
+            table_spec.as_ref().unwrap().clone(),
```
nit: justify unwrap

Lorak-mmk replied:
+1, it's not obvious to me that `col_count` must be greater than 0.

wprzytula replied:
Good that you noticed that bad practice. I refactored the code much.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -542,87 +595,95 @@ fn deser_table_spec(buf: &mut &[u8]) -> StdResult<TableSpec<'static>, TableSpecP
+                .map_err(|err| CqlTypeParseError::TypeIdParseError(err.into()))?;
+            Ok(match id {
+                0x0000 => {
+                    let type_str = types::read_string(buf)
```
This is done on purpose. Chances are the underlying string is ...DurationType, in which case we don't need to allocate it at all. Only for Custom types (which we don't support anyway) do we need to allocate. OTOH, the macro argument function deserializes borrowed OR owned string; here we want to always deserialize borrowed string.

muzarski replied:
> This is done on purpose. Chances are the underlying string is ...DurationType,

Right, this makes sense to me now

Lorak-mmk commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -513,13 +513,14 @@ impl From<SerializationError> for PartitionKeyError {
-pub(crate) type PartitionKeyValue<'ps> = (&'ps [u8], &'ps ColumnSpec);
+pub(crate) type PartitionKeyValue<'ps, 'spec> = (&'ps [u8], &'ps ColumnSpec<'spec>);
 
-pub(crate) struct PartitionKey<'ps> {
-    pk_values: SmallVec<[Option<PartitionKeyValue<'ps>>; PartitionKey::SMALLVEC_ON_STACK_SIZE]>,
+pub(crate) struct PartitionKey<'ps, 'spec: 'ps> {
+    pk_values:
+        SmallVec<[Option<PartitionKeyValue<'ps, 'spec>>; PartitionKey::SMALLVEC_ON_STACK_SIZE]>,
```
Is there a reason for having 2 lifetimes? I think this type is covariant over both 'ps and 'spec, so it should be fine to unify those lifetimes - but maybe I'm just talking gibberish, idk. @piodul ?

Lorak-mmk commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -558,6 +558,16 @@ impl ResultMetadata {
+    pub fn col_count(&self) -> usize {
+        self.col_count
+    }
+
+    #[inline]
+    pub fn col_specs(&self) -> &[ColumnSpec] {
+        &self.col_specs
+    }
```
Could you use `ColumnSpec<'_>` to signify that lifetime is not static?

wprzytula replied:
Actually, what lifetime does the current syntax imply? How does it differ from `'_` ?

Lorak-mmk replied:
Current syntax does imply `'_` lifetime - I did not ask for a change in semantics, but for highlighting current semantics.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -61,23 +61,72 @@ pub enum ColumnType {
+            ColumnType::Date => ColumnType::Date,
+            ColumnType::Decimal => ColumnType::Decimal,
+            ColumnType::Double => ColumnType::Double,
+            ColumnType::Duration => ColumnType::Duration,
+            ColumnType::Float => ColumnType::Float,
+            ColumnType::Int => ColumnType::Int,
+            ColumnType::BigInt => ColumnType::BigInt,
+            ColumnType::Text => ColumnType::Text,
+            ColumnType::Timestamp => ColumnType::Timestamp,
+            ColumnType::Inet => ColumnType::Inet,
+            ColumnType::List(elem_type) => ColumnType::List(Box::new(elem_type.into_owned())),
+            ColumnType::Map(key_type, value_type) => ColumnType::Map(
+                Box::new(key_type.into_owned()),
+                Box::new(value_type.into_owned()),
+            ),
+            ColumnType::Set(elem_type) => ColumnType::Set(Box::new(elem_type.into_owned())),
+            ColumnType::UserDefinedType {
+                type_name,
+                keyspace,
+                field_types,
+            } => ColumnType::UserDefinedType {
+                type_name: type_name.into_owned().into(),
+                keyspace: keyspace.into_owned().into(),
+                field_types: field_types
+                    .into_iter()
+                    .map(|(cow, column_type)| (cow.into_owned().into(), column_type.into_owned()))
+                    .collect(),
+            },
+            ColumnType::SmallInt => ColumnType::SmallInt,
+            ColumnType::TinyInt => ColumnType::TinyInt,
+            ColumnType::Time => ColumnType::Time,
+            ColumnType::Timeuuid => ColumnType::Timeuuid,
+            ColumnType::Tuple(vec) => {
+                ColumnType::Tuple(vec.into_iter().map(ColumnType::into_owned).collect())
+            }
+            ColumnType::Uuid => ColumnType::Uuid,
+            ColumnType::Varint => ColumnType::Varint,
+        }
+    }
+}
```
AFAIK this function is only going to be used in error conditions, to create owned `ColumnType` for use in error types, so the waste does not matter.

wprzytula commented on scylla-cql/src/frame/value_tests.rs:
```diff
@@ -1079,7 +1079,7 @@ fn vec_value_list() {
-fn col_spec(name: &str, typ: ColumnType) -> ColumnSpec {
+fn col_spec(name: &str, typ: ColumnType<'static>) -> ColumnSpec {
     ColumnSpec {
         table_spec: TableSpec::owned("ks".to_string(), "tbl".to_string()),
```
This could be done with any lifetime-generic type that we have in our codebase, and there are quite a number of them. Should we introduce such an alias for each of them?
I'm a bit reluctant about type aliases, because they tend to hide that the underlying type is the same **yet** differs wrt ownership. The thing is, we want to expose that fact.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -698,7 +722,7 @@ fn deser_col_specs(
+) -> StdResult<Vec<ColumnSpec<'static>>, ColumnSpecParseError> {
     let mut col_specs = Vec::with_capacity(col_count);
     for col_idx in 0..col_count {
         let table_spec = if let Some(spec) = global_table_spec {
```
(already answered elsewhere)

wprzytula commented on scylla-cql/src/types/deserialize/value.rs:
```diff
@@ -1146,16 +1147,16 @@ impl_tuple_multiple!(
 /// - `Some(None)` - present, but null
 /// - `Some(Some(...))` - non-null, present value
 pub struct UdtIterator<'frame> {
-    all_fields: &'frame [(String, ColumnType)],
+    all_fields: &'frame [(Cow<'frame, str>, ColumnType<'frame>)],
     type_name: &'frame str,
     keyspace: &'frame str,
-    remaining_fields: &'frame [(String, ColumnType)],
+    remaining_fields: &'frame [(Cow<'frame, str>, ColumnType<'frame>)],
     raw_iter: BytesSequenceIterator<'frame>,
 }
 
```
This is a valid question. `ColumnType::UserDefinedType` contains `Vec<(Cow<'frame, str>, ColumnType<'frame>)>`, so when borrowing from that Vec, we get `&'frame [(Cow<'frame, str>, ColumnType<'frame>)]`, which can't be cast to `[(&'frame str, ColumnType<'frame>)]. So, the answer is that `Cow`s have to be propagated.

wprzytula commented on scylla-cql/src/frame/response/result.rs:
```diff
@@ -424,9 +424,38 @@ impl CqlValue {
+    pub fn table_spec(&self) -> &TableSpec {
+        &self.table_spec
+    }
+
```
This is a bug. Thanks for noticing that. It should be `TableSpec<'static>`.

wprzytula commented:
v2.0:
- refactored `deser_table_spec`/`deser_col_specs` tandem again to remove `unwrap()` and for more sense in responsibility division,
- addressed comments.

wprzytula commented:
In the PR description, I listed out the changes introduced.
